EventFilter/HcalRawToDigi/plugins/HcalDigiToRawuHTR.cc

```diff
@@ -0,0 +1,282 @@
+#include "DataFormats/HcalDigi/interface/HcalQIESample.h"
+
+#include "CondFormats/HcalObjects/interface/HcalElectronicsMap.h"
+
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/FEDRawData/interface/FEDRawDataCollection.h"
+#include "DataFormats/FEDRawData/interface/FEDHeader.h"
+#include "DataFormats/FEDRawData/interface/FEDTrailer.h"
+#include "DataFormats/FEDRawData/interface/FEDNumbering.h"
+#include "DataFormats/FEDRawData/interface/FEDRawData.h"
+#include "FWCore/Utilities/interface/CRC16.h"
+
+#include "DataFormats/HcalDigi/interface/HcalDigiCollections.h"
+#include "CalibFormats/HcalObjects/interface/HcalDbService.h"
+#include "CalibFormats/HcalObjects/interface/HcalDbRecord.h"
+
+#include "PackerHelp.h"
+
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+
+/* QUESTION: what about dual FED readout? */
+/* QUESTION: what do I do if the number of 16-bit words
+   are not divisible by 4? -- these need to
+   fit into the 64-bit words of the FEDRawDataFormat */
+
+
+using namespace std;
+
+class HcalDigiToRawuHTR : public edm::stream::EDProducer<> {
+public:
+  explicit HcalDigiToRawuHTR(const edm::ParameterSet&);
+  ~HcalDigiToRawuHTR();
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  virtual void produce(edm::Event&, const edm::EventSetup&) override;
+  void getData(const edm::Event&, const edm::EventSetup&);
+
+  int _verbosity;
+  std::string electronicsMapLabel_;
+
+  edm::EDGetTokenT<HcalDataFrameContainer<QIE10DataFrame> > tok_QIE10DigiCollection_;
+  edm::Handle<QIE10DigiCollection> qie10DigiCollection;
+  edm::EDGetTokenT<HcalDataFrameContainer<QIE11DataFrame> > tok_QIE11DigiCollection_;
+  edm::Handle<QIE11DigiCollection> qie11DigiCollection;
+  edm::EDGetTokenT<HBHEDigiCollection> tok_HBHEDigiCollection_;
+  edm::Handle<HBHEDigiCollection> hbheDigiCollection;
+  edm::EDGetTokenT<HFDigiCollection> tok_HFDigiCollection_;
+  edm::Handle<HFDigiCollection> hfDigiCollection;
+  edm::EDGetTokenT<HcalTrigPrimDigiCollection> tok_TPDigiCollection_;
+  edm::Handle<HcalTrigPrimDigiCollection> tpDigiCollection;
+
+  edm::InputTag qie10Tag_, qie11Tag_, hbheqie8Tag_, hfqie8Tag_, trigTag_;
+};
+
+HcalDigiToRawuHTR::HcalDigiToRawuHTR(const edm::ParameterSet& iConfig) :
+  _verbosity(iConfig.getUntrackedParameter<int>("Verbosity", 0)),
+  electronicsMapLabel_(iConfig.getParameter<std::string>("ElectronicsMap")),
+  qie10Tag_(iConfig.getUntrackedParameter<edm::InputTag>("QIE10", edm::InputTag("hcalDigis"))),
+  qie11Tag_(iConfig.getUntrackedParameter<edm::InputTag>("QIE11", edm::InputTag("hcalDigis"))),
+  hbheqie8Tag_(iConfig.getUntrackedParameter<edm::InputTag>("HBHEqie8", edm::InputTag("hcalDigis"))),
+  hfqie8Tag_(iConfig.getUntrackedParameter<edm::InputTag>("HFqie8", edm::InputTag("hcalDigis"))),
+  trigTag_(iConfig.getUntrackedParameter<edm::InputTag>("TP", edm::InputTag("hcalDigis")))
```

Comment:
all parameters that affect the persisted outputs should be tracked.
"Verbosity" is the only good untracked parameter.